### ISSUE TITLE
use the same data but get different result

### ISSUE THREAD

**Issue by haha010508** [YingfanWang/PaCMAP#3]
i use same data (dim = 512), but i get different result, this is my code:
`embedding_pacmap = pacmap.PaCMAP(n_dims=2, n_neighbors=None, MN_ratio=0.5, FP_ratio=2.0)
 X_transformed = embedding_pacmap.fit_transform(partxvec, init="pca")`

and at the same time, i use UMAP, i set random papramater
`   reducer = umap.UMAP(random_state=42)
   embedding = reducer.fit_transform(partxvec)`
and i get the same result,
so i want to know whether this algorithm have same configer with UMAP?

**Comment by hyhuang00** [comment]
It's hard for me to to tell without knowing which dataset you are looking at and how different the result is. Could you elaborate on that?
To replicate the result of the figure in this repo, we suggest you to use n_neighbors=10 for PaCMAP.

**Comment by pmbaumgartner** [comment]
@hyhuang00 I think the question was this: is there an equivalent `random_state` parameter so that we can exactly replicate results?

If you run the experiments I think you'll get similar, but not exactly the same results. There are several places in the algorithm where it looks like a random number is selected (e.g. https://github.com/YingfanWang/PaCMAP/blob/master/pacmap/pacmap.py#L90) that would make this non-deterministic.

Having exact replicability would be a great benefit for this library.

**Comment by hyhuang00** [comment]
> @hyhuang00 I think the question was this: is there an equivalent `random_state` parameter so that we can exactly replicate results?
> 
> If you run the experiments I think you'll get similar, but not exactly the same results. There are several places in the algorithm where it looks like a random number is selected (e.g. https://github.com/YingfanWang/PaCMAP/blob/master/pacmap/pacmap.py#L90) that would make this non-deterministic.
> 
> Having exact replicability would be a great benefit for this library.

That's a good point! We will work on that to make sure every random number can be controlled by a single random seeding mechanism.

**Comment by pmbaumgartner** [comment]
Glad to hear it! PACMAP is working great for me so far, and having deterministic results will make it even more useful. Thank you!

**Comment by robvanvolt** [comment]
> > @hyhuang00 I think the question was this: is there an equivalent `random_state` parameter so that we can exactly replicate results?
> > If you run the experiments I think you'll get similar, but not exactly the same results. There are several places in the algorithm where it looks like a random number is selected (e.g. https://github.com/YingfanWang/PaCMAP/blob/master/pacmap/pacmap.py#L90) that would make this non-deterministic.
> > Having exact replicability would be a great benefit for this library.
> 
> That's a good point! We will work on that to make sure every random number can be controlled by a single random seeding mechanism.

Any news on the random_seed?:)

**Comment by hyhuang00** [comment]
Not much :( We have encountered some difficulties in handling the interaction between random seeding and numba, and we are still working on solving this issue.

**Comment by hyhuang00** [comment]
> > > @hyhuang00 I think the question was this: is there an equivalent `random_state` parameter so that we can exactly replicate results?
> > > If you run the experiments I think you'll get similar, but not exactly the same results. There are several places in the algorithm where it looks like a random number is selected (e.g. https://github.com/YingfanWang/PaCMAP/blob/master/pacmap/pacmap.py#L90) that would make this non-deterministic.
> > > Having exact replicability would be a great benefit for this library.
> > 
> > 
> > That's a good point! We will work on that to make sure every random number can be controlled by a single random seeding mechanism.
> 
> Any news on the random_seed?:)

v0.5.0 has just been released -- I tried my best to fix every place where randomness is introduced and in this new version you will be able to control all the seeds involved using one parameter. Unfortunately, the result can still be indeterministic due to numba parallelization, and fixing the problem requires a full re-implementation of some of the core modules inside the algorithm. The good news is because PaCMAP is pretty robust, given a fixed init and set of pairs the result should be comparable. I will think about ways to deal with the parallelization.
